File: problems/844_backspace_string_compare.rb
=begin
Given two strings S and T, return if they are equal when both are typed into empty text editors. # means a backspace character.

Example 1:

Input: S = "ab#c", T = "ad#c"
Output: true
Explanation: Both S and T become "ac".
Example 2:

Input: S = "ab##", T = "c#d#"
Output: true
Explanation: Both S and T become "".
Example 3:

Input: S = "a##c", T = "#a#c"
Output: true
Explanation: Both S and T become "c".
Example 4:

Input: S = "a#c", T = "b"
Output: false
Explanation: S becomes "c" while T becomes "b".
Note:

1 <= S.length <= 200
1 <= T.length <= 200
S and T only contain lowercase letters and '#' characters.
Follow up:

Can you solve it in O(N) time and O(1) space?
=end

# @param {String} s
# @param {String} t
# @return {Boolean}
def backspace_compare(s, t)
    i = s.length - 1
    j = t.length - 1
    skip_s = 0
    skip_t = 0
    while i >= 0 || j >= 0
        while i >= 0
            if s[i] == '#'
                skip_s += 1
                i -= 1
            elsif skip_s > 0
                skip_s -= 1
                i -= 1
            else
                break
            end
        end
        while j >= 0
            if t[j] == '#'
                skip_t += 1
                j -= 1
            elsif skip_t > 0
                skip_t -= 1
                j -= 1
            else
                break
            end
        end
        
        if i >= 0 && j >=0 && s[i] != t[j]
            return false
        end
        if i >= 0 != j >= 0
            return false
        end
        i -= 1
        j -= 1
    end
    
    return true
end

=begin
O(n+m) space solution
def backspace_compare(s, t)
    build(s) == build(t)
end

def build(string)
   ans = []
    string.each_char do |char|
        if char != '#'
            ans.push(char)
        elsif ans[0]
            ans.pop
        end
    end
    ans.join
end
=end

File: problems/844_backspace_string_compare.rs
/*
 * Given two strings s and t, return true if they are equal when both are typed into empty text editors. '#' means a backspace character.

Note that after backspacing an empty text, the text will continue empty.



Example 1:

Input: s = "ab#c", t = "ad#c"
Output: true
Explanation: Both s and t become "ac".
Example 2:

Input: s = "ab##", t = "c#d#"
Output: true
Explanation: Both s and t become "".
Example 3:

Input: s = "a#c", t = "b"
Output: false
Explanation: s becomes "c" while t becomes "b".


Constraints:

1 <= s.length, t.length <= 200
s and t only contain lowercase letters and '#' characters.


Follow up: Can you solve it in O(n) time and O(1) space?
*/

impl Solution {
    pub fn backspace_compare(s: String, t: String) -> bool {
        let mut a = vec![];
        for c in s.bytes() {
            if c == b'#' {
                a.pop();
            } else {
                a.push(c);
            }
        }
        let mut b = vec![];
        for c in t.bytes() {
            if c == b'#' {
                b.pop();
            } else {
                b.push(c);
            }
        }
        a == b
    }
}

